File: solutions/1305/1305-kir3i.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getAllElements(self, root1, root2):
        ans = []
        t1 = []
        t2 = []
        self.popAscList(root1, t1)
        self.popAscList(root2, t2)
        
        i1 = i2 = 0

        while i1<len(t1) and i2<len(t2):
            if t1[i1] <= t2[i2]:
                ans.append(t1[i1])
                i1 += 1
            else:
                ans.append(t2[i2])
                i2 += 1
        if i1 == len(t1):
            ans.extend(t2[i2:])
        else:
            ans.extend(t1[i1:])
        return ans
    
    def popAscList(self, root, res):
        if not root:
            res = []
            return
        
        if root.left:
            self.popAscList(root.left, res)
        res.append(root.val)
        if root.right:
            self.popAscList(root.right, res)

File: solutions/1305/1305-kir3i.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> ans, t1, t2;
    vector<int> getAllElements(TreeNode* root1, TreeNode* root2) {
        int i1=0, i2=0;
        popAscList(root1, t1);
        popAscList(root2, t2);
        
        while(i1<t1.size() && i2<t2.size()) {
            if (t1[i1] <= t2[i2])
                ans.push_back(t1[i1++]);
            else
                ans.push_back(t2[i2++]);
        }
        
        if (i1 == t1.size())
            ans.insert(ans.end(), t2.begin()+i2, t2.end());
        else
            ans.insert(ans.end(), t1.begin()+i1, t1.end());
        
        return ans;
    }
    
    void popAscList(TreeNode* root, vector<int> &t) {
        if (!root) {
            t = {};
            return;
        }
        
        if (root->left)
            popAscList(root->left, t);
        t.push_back(root->val);
        if (root->right)
            popAscList(root->right, t);
    }
};